### PR TITLE
sold: init at 0-unstable-2024-02-22

### PR DESCRIPTION
**Pull request by EliseZeroTwo**
## Description of changes

Adding a package for sold, a fork of mold that supports linking for iOS & MacOS platforms.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/so/sold/package.nix [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+}:
+
+stdenv.mkDerivation rec {
+  pname = "sold";
+  version = "1.11.0";
```
Comment (applePrincess) [inline pkgs/by-name/so/sold/package.nix:10]: As per https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-naming
The upstream has not made any releases or attached tags, so this should be
```suggestion
  version = "0-unstable-2024-02-22";
```

File: pkgs/by-name/so/sold/package.nix [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+  nativeBuildInputs = [ cmake ];
+  enableParallelBuilding = true;
+
+  meta = {
+    description = "A fork of mold, a faster drop-in replacement for existing Unix linkers, with support for iOS and MacOS";
+    license = lib.licenses.mit;
+    homepage = "https://github.com/bluewhalesystems/sold";
+    maintainers = with lib.maintainers; [ elise ];
+    platforms = lib.platforms.all;
```
Comment (applePrincess) [inline pkgs/by-name/so/sold/package.nix:27]: `mainProgram` is missing.

**Review by applePrincess — CHANGES_REQUESTED** [review]

File: pkgs/by-name/so/sold/package.nix (hunk 1 above)
Comment (EliseZeroTwo) [reply pkgs/by-name/so/sold/package.nix]: Thank you, I have resolved this

File: pkgs/by-name/so/sold/package.nix (hunk 2 above)
Comment (EliseZeroTwo) [reply pkgs/by-name/so/sold/package.nix:27]: Thank you, I think I have resolved this, I placed what I feel would be the most appropriate binary there

File: pkgs/by-name/so/sold/package.nix [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  src = fetchFromGitHub {
+    owner = "bluewhalesystems";
+    repo = pname;
+    rev = "a52e8b37661562c866a579ab1164f94694f479a5";
+    sha256 = "0iyrgz0ay2453rqrlvxsrla0998l8yq6q3xyqqw5icfvh05ld8i8";
+  };
+
+  nativeBuildInputs = [ cmake ];
+  enableParallelBuilding = true;
+
```
Comment (applePrincess) [inline pkgs/by-name/so/sold/package.nix:21]: openssl seems to be required at build time and run time (at least on x86_64-linux)
so

```nix
  buildInputs = [ openssl ];
```

**Review by applePrincess — COMMENTED** [review]
Sorry for another change request

**Review by Kamillaova — CHANGES_REQUESTED** [review]
I think you forgot to add this package to all-packages.nix like `mold`, and `sold-wrapper` like `mold-wrapped`

And referring to `mold` this package should be located at pkgs/development/tools/sold

**Comment by marcusramberg** [comment]
> I think you forgot to add this package to all-packages.nix like `mold`, and `sold-wrapper` like `mold-wrapped`
> 
> And referring to `mold` this package should be located at pkgs/development/tools/sold

That isn't necessary as it's added in the new by-name tree.

**Comment by Kamillaova** [comment]
> That isn't necessary as it's added in the new by-name tree.

image

File: pkgs/by-name/so/sold/package.nix (hunk 3 above)
Comment (EliseZeroTwo) [reply pkgs/by-name/so/sold/package.nix:23]: Done, thank you!

File: pkgs/by-name/so/sold/package.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  openssl,
+}:
+
+stdenv.mkDerivation rec {
+  pname = "sold";
+  version = "0-unstable-2024-02-22";
```
[the remark this answers is not in the document]
Comment (EliseZeroTwo) [reply pkgs/by-name/so/sold/package.nix:11]: https://github.com/NixOS/nixpkgs/pull/297741#discussion_r1534170996

[the remark this answers is not in the document]
Comment (Kamillaova) [reply pkgs/by-name/so/sold/package.nix:11]: Maybe rename the commit and pull request?

**Review by lolbinarycat — APPROVED** [review]
built on x86_64-linux

File: pkgs/by-name/so/sold/package.nix [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  src = fetchFromGitHub {
+    owner = "bluewhalesystems";
+    repo = pname;
```
Comment (Aleksanaa) [inline pkgs/by-name/so/sold/package.nix:15]:
```suggestion
    repo = "sold";
```

File: pkgs/by-name/so/sold/package.nix [hunk 6]
```diff
@@ -0,0 +1,32 @@
+{
+  lib,
+  stdenv,
+  fetchFromGitHub,
+  cmake,
+  openssl,
+}:
+
+stdenv.mkDerivation rec {
```
Comment (Aleksanaa) [inline pkgs/by-name/so/sold/package.nix:9]:
```suggestion
stdenv.mkDerivation {
```

**Comment by pluiedev** [comment]
Seems that https://github.com/bluewhalesystems/sold had been archived & deprecated a few days ago in favor of Apple's new linker (included in Xcode 15 and newer)

**Pull request closed without merging**